src/File.jl

```diff
@@ -217,3 +217,112 @@ function get_files(
     end
     return result
 end
+
+"""
+    function get_sub(
+        path;
+        from_fname::Bool=true,
+        require_modality::Bool=true,
+        strict::Bool=true
+    )
+
+Function to get subject_id from path or File object.
+
+The following keyword arguments can be passed:
+
+* `from_fname`::Bool - detauls to true. If `true`, only looks the `subject_id` from
+  filename, otherwise, looks into full path if `subject_id` can't be found in filename
+* `require_modality::Bool` - defaults to `true`. If `true` this expects a modality
+  presents in every file name, e.g. sub-subtest_ses-1_run-001_**T1w**.nii.gz
+* `strict::Bool` - defaults to `true`. If `true`, `BIDSTools` will throw an error on
+  invalid BIDS filename, this can be turned to `false` to not parse those file names
+  and display a warning instead. This will result in empty dictionary in `entities`.
+
+Returns nothing if no subject ID found.
+"""
+get_sub(file::File; kws...) = get_sub(file.path; kws...)
+
+function get_sub(
+    path::AbstractString;
+    from_fname::Bool=true,
+    require_modality::Bool=true,
+    strict::Bool=true
+)
+    sub_rgx =  r"[\\|/]sub-(.+?)[\\|/]"
+    sub_match = get(
+        parse_fname(basename(path), require_modality=require_modality, strict=strict),
+        "sub",
+        nothing
+    )
+    if isnothing(sub_match) && !from_fname
+        sub_match = isnothing(match(sub_rgx, path)) ? nothing : match(sub_rgx, path)[1]
+    end
+    !isnothing(sub_match) && return sub_match
+    nothing
+end
+
+"""
+    function get_ses(
+        path;
+        from_fname::Bool=true,
+        require_modality::Bool=true,
+        strict::Bool=true
+    )
+
+Function to get session_id from path or File object.
+
+The following keyword arguments can be passed:
+
+* `from_fname`::Bool - detauls to true. If `true`, only looks the `session_id` from
+  filename, otherwise, looks into full path if `session_id` can't be found in filename
+* `require_modality::Bool` - defaults to `true`. If `true` this expects a modality
+  presents in every file name, e.g. sub-subtest_ses-1_run-001_**T1w**.nii.gz
+* `strict::Bool` - defaults to `true`. If `true`, `BIDSTools` will throw an error on
+  invalid BIDS filename, this can be turned to `false` to not parse those file names
+  and display a warning instead. This will result in empty dictionary in `entities`.
+
+Returns nothing if no session ID found.
+"""
+get_ses(file::File; kws...) = get_ses(file.path; kws...)
+
+function get_ses(
+    path::AbstractString;
+    from_fname::Bool=true,
+    require_modality::Bool=true,
+    strict::Bool=true
+)
+    ses_rgx =  r"[\\|/]ses-(.+?)[\\|/]"
```

Comment:
Thanks! Yes you're right on the pipe `[\\/]` is exactly what I need.

In regards to the `?`, I use it to capture the shortest match since I'd like to extract these from the full path, i.e. 

```julia
julia> ses_rgx =  r"[\\/]ses-(.+)[\\/]"
r"[\\/]ses-(.+)[\\/]"

julia> match(ses_rgx, "root/sub-test/ses-1/anat/run-001.nii.gz")
RegexMatch("/ses-1/anat/", 1="1/anat")

julia> ses_rgx =  r"[\\/]ses-(.+?)[\\/]"
r"[\\/]ses-(.+?)[\\/]"

julia> match(ses_rgx, "root/sub-test/ses-1/anat/run-001.nii.gz")
RegexMatch("/ses-1/", 1="1")
```

Reply:
Oh right, very good.

I guess I would have done that by excluding the path separator from the session match:

```
ses_rgx =  r"[\\/]ses-([^\\/]+)[\\/]"
```

though on further thought, I think what you have should be about as efficient, and possibly more readable. (Provided you don't forget that shortest match syntax is a thing like I did!)